### PR TITLE
Pytorch: add inmemory batched dataloader

### PR DESCRIPTION
**Pull request by chongxiaoc**
Including two changes:
- Remove current inmem_cache option for pytorch batched dataloader
- Create a new class for inmemory batched dataloader

Related to issue#664

### CODE REVIEW

File: petastorm/pytorch.py [hunk 1]
```diff
@@ -367,3 +367,123 @@ def __enter__(self):
...
+            n_rows = expected_rows
+
+        # At this point, dataloader has enough rows storted in memory.
+        # Set last_row_consumed = True in reader rather than draining the remainder
+        # of reader. If reader has infinite epochs, draining will be deadlock.
+        for row in self.reader:
+            self.reader.last_row_consumed = True
+        # Truncate empty tensors if capacity is larger than total rows.
+        if n_rows < self.capacity:
+            for k, v in enumerate(self.buffer):
```
Comment (selitvin) [inline petastorm/pytorch.py:455]: nit: Perhaps `for i, v in ...` to help reader understand that this is an iteration number?

File: petastorm/pytorch.py [hunk 2]
```diff
@@ -367,3 +367,123 @@ def __enter__(self):
...
+            # Copy new items into buffer
+            for k, v in enumerate(items):
+                self.buffer[k][n_rows:expected_rows] = v[:last_row]
+            n_rows = expected_rows
+
+        # At this point, dataloader has enough rows storted in memory.
+        # Set last_row_consumed = True in reader rather than draining the remainder
+        # of reader. If reader has infinite epochs, draining will be deadlock.
+        for row in self.reader:
+            self.reader.last_row_consumed = True
```
Comment (selitvin) [inline petastorm/pytorch.py:452]: I think what you want to do instead of this `for` loop is `self.reader.stop()`.

File: petastorm/pytorch.py [hunk 3]
```diff
@@ -367,3 +367,123 @@ def __enter__(self):
...
+
+            if self.capacity <= expected_rows:
+                buffer_full = True
+                last_row = self.capacity-n_rows
+                expected_rows = self.capacity
+            if self.buffer is None:
+                # Initialize buffer as a list of empty tensors
+                self.buffer = []
+                for v in items:
+                    self.buffer.append(torch.empty((self.capacity,) + v.shape[1:], dtype=v.dtype, device=v.device))
```
Comment (selitvin) [inline petastorm/pytorch.py:442]: Can you please clarify what does the `+v.shape[1:]` term purpose?

File: petastorm/pytorch.py [hunk 4]
```diff
@@ -367,3 +367,123 @@ def __enter__(self):
...
+
+    def _load_rows_into_mem(self):
+        n_rows = 0
+        buffer_full = False
+        self.buffer = None
+        self.keys = None
+
+        for row in self.reader:
+            if buffer_full:
+                break
```
Comment (selitvin) [inline petastorm/pytorch.py:412]: It's easy for a user to assume that if they use this class all data going to be loaded into memory. They might not notice the `capacity` constructor argument since it has a default value. I think we should make `capacity` a mandatory argument or find some other API change that make this behavior of dropping rows more obvious. 

Perhaps tweaking the name to hint that the value is given as rows (and not bytes), e.g. rows_capacity.

File: petastorm/pytorch.py [hunk 5]
```diff
@@ -367,3 +367,123 @@ def __enter__(self):
...
+                self.buffer[k] = self.buffer[k][:n_rows]
+
+    def _iter_impl(self):
+        """
+        The Data Loader iterator stops the for-loop when num_epochs is reached.
+        """
+        self._load_rows_into_mem()
+
+        for epoch in range(self.num_epochs):
+            self.size = len(self.buffer[0])
```
Comment (selitvin) [inline petastorm/pytorch.py:465]: `size` can be a local variable. No reason for it to be the class state from what I see.

File: petastorm/pytorch.py [hunk 6]
```diff
@@ -367,3 +367,123 @@ def __enter__(self):
...
+                batch = dict(zip(self.keys, batch))
+                yield batch
+            assert self.size == 0
+
+    # Functions needed to treat data loader as a context manager
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.reader.stop()
```
Comment (selitvin) [inline petastorm/pytorch.py:488]: Your logic shuts the reader earlier - once all data was read. It's a good thing since we release resources we would no longer need. No need to `stop()` the reader here, but ok to join to make sure the `InMemBatchedDataLoader` indeed acts as  a context-manager and does not exit until everything is clean.

File: petastorm/pytorch.py [hunk 7]
```diff
@@ -367,3 +367,123 @@ def __enter__(self):
...
+        self.reader = reader
+        self.batch_size = batch_size
+        self.transform_fn = transform_fn or torch.as_tensor
+        self.num_epochs = num_epochs
+        self.seed = seed
+        self.capacity = capacity
+        self.shuffle = shuffle
+        self.size = 0
+
+    def _load_rows_into_mem(self):
```
Comment (selitvin) [inline petastorm/pytorch.py:404]: Would making this a standalone (non class function) that returns `(keys, buffers) ` tuple more clear and easier to reuse if needed?
```python
_loader_rows_into_mem(reader, transform_fn, capacity)
```
We would need to `stop()` the reader in iter_impl though.

Having buffer and keys part of the class state is unnecessary since it is not used after `iter_impl` exists (hence it would be good to release the memory). Or am I wrong and `InMemBatchedDataLoader`  could be used multiple times somehow?

File: petastorm/pytorch.py [hunk 8]
```diff
@@ -367,3 +367,123 @@ def __enter__(self):
...
+        # Truncate empty tensors if capacity is larger than total rows.
+        if n_rows < self.capacity:
+            for k, v in enumerate(self.buffer):
+                self.buffer[k] = self.buffer[k][:n_rows]
+
+    def _iter_impl(self):
+        """
+        The Data Loader iterator stops the for-loop when num_epochs is reached.
+        """
+        self._load_rows_into_mem()
```
Comment (selitvin) [inline petastorm/pytorch.py:462]: How about:
```python

if self.buffer is not None:
  buffer, keys = _load_rows_into_mem(...)
```
this way, we:
 1. don't need to reread the data as in `test_call_iter_on_dataloader_multiple_times_inmem` test case
 2. can safely stop the read after the loading of the data finished first time.

File: petastorm/tests/test_pytorch_dataloader.py [hunk 9]
```diff
@@ -208,3 +256,68 @@ def test_call_iter_on_dataloader_multiple_times(synthetic_dataset, reader_factor
...
+
+@pytest.mark.parametrize('reader_factory', MINIMAL_READER_FLAVOR_FACTORIES)
+def test_inmem_batched_dataloader_shuffle_per_epoch(synthetic_dataset, reader_factory):
+    """See if we are getting different shuffled result in different epochs"""
+    epochs = 3
+    with InMemBatchedDataLoader(reader_factory(synthetic_dataset.url, schema_fields=['^id$']),
+                                batch_size=10, capacity=100, num_epochs=epochs, shuffle=True) as loader:
+        first_result = []
+        second_result = []
+        third_result = []
```
Comment (selitvin) [inline petastorm/tests/test_pytorch_dataloader.py:301]: Unused variable

File: petastorm/pytorch.py [hunk 10]
```diff
@@ -367,3 +367,123 @@ def __enter__(self):
...
+        :param capacity: number of rows to be loaded into memory.
+        :param shuffle: If ``True``, indices will be shuffled in every epoch.
+        """
+        super(InMemBatchedDataLoader, self).__init__()
+        self.reader = reader
+        self.batch_size = batch_size
+        self.transform_fn = transform_fn or torch.as_tensor
+        self.num_epochs = num_epochs
+        self.seed = seed
+        self.capacity = capacity
```
[the remark this answers is not in the document]
Comment (chongxiaoc) [reply petastorm/pytorch.py:400]: sure thing, I was following existing style. let me put them all private.

File: petastorm/pytorch.py (hunk 3 above)
Comment (chongxiaoc) [reply petastorm/pytorch.py:442]: We want to allocate empty tensor with exact dimension of input rows.
v.shape[0] is row group size, while v.shape[1:] is possible dimensions of input row data.
This is same as code in shuffling buffer:  https://github.com/uber/petastorm/blob/7f37e8dde6ff1b13f055d22a6289e2de8bb5d473/petastorm/reader_impl/pytorch_shuffling_buffer.py#L269

**Comment by chongxiaoc** [comment]
TODO: mention the class in the main Readme and release notes

File: petastorm/pytorch.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (chongxiaoc) [reply petastorm/pytorch.py:400]: It turns out that I have to set `reader` is public attribute, since inmem loader class inherit from `LoaderBase`, which users `.reader`.

[the remark this answers is not in the document]
Comment (chongxiaoc) [reply petastorm/pytorch.py:400]: Well, if `InmemDataLoader` doesn't support multiple iterations, it doesn't have to inherit `LoderBase`.

**Comment by chongxiaoc** [comment]
@selitvin ready to be reviewed again.

File: petastorm/pytorch.py [hunk 11]
```diff
@@ -423,3 +367,139 @@ def __enter__(self):
...
+    def __init__(self, reader, batch_size=1,
+                 transform_fn=None,
+                 num_epochs=1,
+                 seed=0,
+                 rows_capacity=1024,
+                 shuffle=False):
+        """
+        Initializes a data loader object.
+        This class does not support special types that are not supported in PyTorch (decimal/string).
+        :param reader: petastorm Reader instance
```
Comment (selitvin) [inline petastorm/pytorch.py:450]: Perhaps mention here that the reader will be `stop`ped once all data is loaded from it.

File: petastorm/pytorch.py [hunk 12]
```diff
@@ -423,3 +367,139 @@ def __enter__(self):
...
+    def __iter__(self):
+        """
+        The Data Loader iterator stops the for-loop when num_epochs is reached.
+        """
+        if self._in_iter:
+            raise RuntimeError("InMemBatchedDataLoader couldn't be used multiple times, please\
+                    specify total number of epochs using num_epochs in constructor.")
+        self._in_iter = True
+        try:
+            keys, buffer = _load_rows_into_mem(self._reader, self._transform_fn, self._rows_capacity)
```
Comment (selitvin) [inline petastorm/pytorch.py:477]: Consider adding a bit more comments describing structure of `keys` and `buffer`. I think it may help a reader understand the code better before digging into `_load_rows_into_mem` implementation.

File: petastorm/pytorch.py [hunk 13]
```diff
@@ -423,3 +367,139 @@ def __enter__(self):
...
+                # Sample batches
+                for i in range(0, len(indices), self._batch_size):
+                    idx = indices[i:i+self._batch_size]
+                    batch = [v[idx] for v in buffer]
+                    size -= len(batch[0])
+                    batch = dict(zip(keys, batch))
+                    yield batch
+                assert size == 0
+        except Exception as e:
+            logger.error('Iteration on InMemBatchedDataLoader raise error: %s', repr(e))
```
Comment (selitvin) [inline petastorm/pytorch.py:497]: What is the added value of this generic exception handler?

File: petastorm/pytorch.py [hunk 14]
```diff
@@ -423,3 +367,139 @@ def __enter__(self):
...
+            raise
+
+    # Functions needed to treat data loader as a context manager
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._reader.join()
```
Comment (selitvin) [inline petastorm/pytorch.py:505]: Would we get into a deadlock if a user instantiated `InMemBatchedDataLoader` but never iterated? Perhaps should unit-test that case.

File: petastorm/tests/test_pytorch_dataloader.py [hunk 15]
```diff
@@ -9,10 +8,12 @@
...
 
 from petastorm import make_reader, TransformSpec, make_batch_reader
-from petastorm.pytorch import _sanitize_pytorch_types, DataLoader, BatchedDataLoader, decimal_friendly_collate
+from petastorm.pytorch import (_sanitize_pytorch_types, DataLoader,
+                               BatchedDataLoader, decimal_friendly_collate,
+                               InMemBatchedDataLoader, _load_rows_into_mem)
 from petastorm.tests.test_common import TestSchema
 
-ALL_DATA_LOADERS = [DataLoader, BatchedDataLoader]
+OOM_DATA_LOADERS = [DataLoader, BatchedDataLoader]
```
Comment (selitvin) [inline petastorm/tests/test_pytorch_dataloader.py:16]: Does OOM stands for out-of-memory? Not sure what would be a better name, but it's a bit strange to see it here :)

File: petastorm/pytorch.py [hunk 16]
```diff
@@ -423,3 +367,139 @@ def __enter__(self):
...
+        self._rows_capacity = rows_capacity
+        self._shuffle = shuffle
+        self._in_iter = False
+
+    def __iter__(self):
+        """
+        The Data Loader iterator stops the for-loop when num_epochs is reached.
+        """
+        if self._in_iter:
+            raise RuntimeError("InMemBatchedDataLoader couldn't be used multiple times, please\
```
Comment (selitvin) [inline petastorm/pytorch.py:473]: If the data loader can not be restarted, then what do you think about a possibility of moving 
```python
self._keys, self._buffer = _load_rows_into_mem(self._reader, self._transform_fn, self._rows_capacity)
```
into the constructor? Then time to the first sample might be improved 
 - for the cases when does something heavy between construction and first read
 - step-through interactive debugging.

File: petastorm/pytorch.py (hunk 13 above)
Comment (chongxiaoc) [reply petastorm/pytorch.py:497]: I'm not sure if `try` block should be used TBH. I was seeing the `LoaderBase` having `try` block for `__iter__`. I feel the only exception may come from `_load_rows_into_mem ` with OOM or reader corruption? Any suggestions?

Comment (selitvin) [reply petastorm/pytorch.py:497]: It could also come from within reader errors. In LoaderBase, this PR: #544 added the try/except and the goal was to raise an explicit error in the second iteration. I think the goal was to support multiple `iter(data_loader)` calls, but fail the next time if previous iterations ended up with an exception.

File: petastorm/tests/test_pytorch_dataloader.py (hunk 15 above)
Comment (chongxiaoc) [reply petastorm/tests/test_pytorch_dataloader.py:16]: Or name them as `BASIC_DATA_LOADERS`?

Comment (selitvin) [reply petastorm/tests/test_pytorch_dataloader.py:16]: That works. Just was surprised to see an OOM here :)

File: petastorm/pytorch.py (hunk 16 above)
Comment (chongxiaoc) [reply petastorm/pytorch.py:473]: sure. I was thinking about this before. Then I though it would not work out with multiple iterations.
Now since we understand that multiple iterations are not allowed, then it should be moved into constructor.

File: petastorm/pytorch.py (hunk 13 above)
Comment (chongxiaoc) [reply petastorm/pytorch.py:497]: Hmm, after a second thought, I decide remove `try` block since `_in_iter` is already checked and raised if possible.

File: petastorm/pytorch.py (hunk 14 above)
Comment (chongxiaoc) [reply petastorm/pytorch.py:505]: - reader is loaded in constructor
- reader is stopped after expected rows are loaded
- work_pool is joined in reader.stop(), here we can leave it nothing just `pass`.
Maybe this is more straightforward?

Comment (chongxiaoc) [reply petastorm/pytorch.py]: I added a timeout test to check possible deadlock.

**Comment by chongxiaoc** [comment]
Addressed new comments. PTAL one more time. @selitvin

File: docs/release-notes.rst [hunk 17]
```diff
@@ -6,6 +6,7 @@ Release notes
 
 Release 0.10.1 (unreleased)
```
Comment (selitvin) [inline docs/release-notes.rst:7]: Need to remember and bump to 0.11.0 since we are changing the interface of `BatchedDataLoader` (removing arguments). No action for this diff.

**Review by selitvin — APPROVED** [review]
Cool stuff! Ship it!

File: docs/release-notes.rst (hunk 17 above)
Comment (chongxiaoc) [reply docs/release-notes.rst:7]: sure thing.
